Edit reprojection and resampling notebook

tcmetzger opened the pull request.

tcmetzger committed:
Edit reprojection and resampling notebook

tcmetzger commented:
This PR contains my edits to the reprojection/resampling notebook. 
cc @marcozimmermannpm

tcmetzger commented:
@thuydotm I changed a lot of the cells to highlight the resampling aspects a little more. I also changed the title of the notebook to "Reprojecting and resampling". Please make sure I didn't add any errors to the whole thing...

A few questions:

- For the introductory paragraph, would it make sense (and be correct) to say "In general, reprojecting and resampling while loading the data is more efficient" (i.e. adding resampling)?
- according to https://github.com/makepath/xarray-spatial/blob/master/xrspatial/multispectral.py, the default values for `true_color()` are `c=10.0` and `th=0.125`. In the notebook, we say that we use `c=20` as a default. Do we set `c` to 20 somewhere, or are we in fact using the default value of 10 instead?
- For "Method 2": Does rasterio.enums.Resampling.average mean the same as https://en.wikipedia.org/wiki/Linear_interpolation? If not, what would be a good resource to link here?
- Also "Method 2": How and where do I specify the target resolution? What default resolution is used if no explicit resolution is provided? Or is the resolution defined in `robinson_transform` (in this case, Step 3 should be (“Calculate a transform to reproject from source CRS to destination CRS and destination resolution.”)?
- For "Method 3": How and where do I set the resolution in this example?
- For "Method 4": Is there a good resource we could link to explain median resampling?
- Also for "Method 4": How and where do I set the target resolution in this example?

 Thanks!

thuydotm commented:
Thanks @tcmetzger. 

Reprojection is actually a resampling operation. When doing reprojection, we resample our data to a new CRS by using an interpolation/resample method. See more here: https://pro.arcgis.com/en/pro-app/2.7/help/analysis/raster-functions/resample-function.htm 

> 1. * For the introductory paragraph, would it make sense (and be correct) to say "In general, reprojecting and resampling while loading the data is more efficient" (i.e. adding resampling)?

Sounds good to me.

> 2. * according to https://github.com/makepath/xarray-spatial/blob/master/xrspatial/multispectral.py, the default values for `true_color()` are `c=10.0` and `th=0.125`. In the notebook, we say that we use `c=20` as a default. Do we set `c` to 20 somewhere, or are we in fact using the default value of 10 instead?

Sorry, I made a mistake here, the default is c=10 and it is used through this notebook.

> 3.  * For "Method 2": Does rasterio.enums.Resampling.average mean the same as https://en.wikipedia.org/wiki/Linear_interpolation? If not, what would be a good resource to link here?

As official `rasterio` documention, average resampling computes the weighted average of all non-NODATA contributing pixels. Maybe we can point user to the "average" part of the section: https://rasterio.readthedocs.io/en/latest/api/rasterio.enums.html#rasterio.enums.Resampling

> 4. * Also "Method 2": How and where do I specify the target resolution? What default resolution is used if no explicit resolution is provided? Or is the resolution defined in `robinson_transform` (in this case, Step 3 should be (“Calculate a transform to reproject from source CRS to destination CRS and destination resolution.”)?

The resolution defined in `robinson_transform`. It is decided by the height, width, and bounds of the destination CRS.

> 5. * For "Method 3": How and where do I set the resolution in this example?

The resolution is set with `space_oblique_resolution` to construct the destination `RasterSpec`

> 6. * For "Method 4": Is there a good resource we could link to explain median resampling?

Similar to 3. I think the best link is the rasterio documentation itself: https://rasterio.readthedocs.io/en/latest/api/rasterio.enums.html#rasterio.enums.Resampling

> 7. * Also for "Method 4": How and where do I set the target resolution in this example?

With rasterio.reproject, the answer is similar to question 4 for method 2.
With rio.reproject, it can be set using `resolution` param. If not provided, it will be calculated internally using information about source raster and destination CRS.

thuydotm commented on tutorials/reprojection_resampling.ipynb:
```diff
@@ -125,12 +125,13 @@
-    " is an equal-area projection that preserves area measurements. Regions with the same size on the Earth's surface have the same size on the map. However, shapes, angles, and scales are generally distorted. "
+    "With `stackstac.stack`, you can use any CRS as defined by an [EPSG code](http://epsg.io/). In this example, you use the `stackstac.stack` function to reproject to the Lambert Cylindrical projection (`epsg:6933`). [Lambert Cylindrical](https://en.wikipedia.org/wiki/Lambert_cylindrical_equal-area_projection) is an equal-area projection that preserves area measurements. Regions with the same si ...
+    "\n",
+    "You also resample the data in this example, using the [bilinear](https://en.wikipedia.org/wiki/Bilinear_interpolation) resampling method. Resampling happens in the `stackstac.stack` function as well. As this function uses [rasterio](https://rasterio.readthedocs.io/en/latest/index.html) internally, you can use any algorithm from the [rasterio.enums.Resampling](https://rasterio.readthedocs.io/ ...
```
We always do resampling when reprojecting, so maybe just say which resampling method we'll be using instead of saying "You also resample the data in this example, .."

tcmetzger committed:
Update text

tcmetzger committed:
Remove empty cell

tcmetzger commented:
@thuydotm Thanks for your input, I have updated the notebook accordingly. For consistency, I divided the code in Method 3 into three steps and added descriptions for each step, like I did for the other methods.

thuydotm commented:
Merged. Thanks so much @tcmetzger!
